### DISCUSSION TITLE
Slow SQL queries cause login times >1 minute

### DISCUSSION THREAD

**Discussion by starcraft66** [photoprism/photoprism#3810] [Q&A]
#### 1. What is not working as documented?

It takes over a minute to log into my photoprism instance.

#### 2. How can we reproduce it?

1. Import hundreds of thousands of files into photoprism
2. Log in
3. Wait

Photoprism reports 232420 items near the "Search" button, all thumbnail, etc settings are at default values.

#### 3. What behavior do you expect?

I expect login to be pretty instantaneous or on the order of a few seconds.

#### 4. What could be the cause of your problem?

Slow SQL queries

#### 5. Can you provide us with example files for testing, error logs, or screenshots?
MariaDB slow query log:

```
/opt/bitnami/mariadb/sbin/mysqld, Version: 11.0.3-MariaDB (Source distribution). started with:
Tcp port: 3306  Unix socket: /opt/bitnami/mariadb/tmp/mysql.sock
Time		   Id Command	Argument
# Time: 231010 19:23:39
# User@Host: root[root] @  [10.233.105.154]
# Thread_id: 1461203  Schema: photoprism  QC_hit: No
# Query_time: 19.053037  Lock_time: 0.000267  Rows_sent: 1  Rows_examined: 927538
# Rows_affected: 0  Bytes_sent: 285
use photoprism;
SET timestamp=1696965819;
SELECT SUM(photo_type = 'video' AND photo_quality > -1 AND photo_private = 0) AS videos, SUM(photo_type = 'live' AND photo_quality > -1 AND photo_private = 0) AS live, SUM(photo_quality = -1) AS hidden, SUM(photo_type IN ('image','animated','vector','raw') AND photo_private = 0 AND photo_quality > -1) AS photos, SUM(photo_type IN ('image','live','animated','vector','raw') AND photo_quality < 3 AND photo_quality > -1 AND photo_private = 0) AS review, SUM(photo_favorite = 1 AND photo_private = 0 AND photo_quality > -1) AS favorites, SUM(photo_private = 1 AND photo_quality > -1) AS private FROM `photos`  WHERE (photos.id NOT IN (SELECT photo_id FROM files WHERE file_primary = 1 AND (file_missing = 1 OR file_error <> ''))) AND (deleted_at IS NULL) LIMIT 1;
# User@Host: root[root] @  [10.233.105.154]
# Thread_id: 1461202  Schema: photoprism  QC_hit: No
# Query_time: 19.120747  Lock_time: 0.000175  Rows_sent: 1  Rows_examined: 927538
# Rows_affected: 0  Bytes_sent: 285
SET timestamp=1696965819;
SELECT SUM(photo_type = 'video' AND photo_quality > -1 AND photo_private = 0) AS videos, SUM(photo_type = 'live' AND photo_quality > -1 AND photo_private = 0) AS live, SUM(photo_quality = -1) AS hidden, SUM(photo_type IN ('image','animated','vector','raw') AND photo_private = 0 AND photo_quality > -1) AS photos, SUM(photo_type IN ('image','live','animated','vector','raw') AND photo_quality < 3 AND photo_quality > -1 AND photo_private = 0) AS review, SUM(photo_favorite = 1 AND photo_private = 0 AND photo_quality > -1) AS favorites, SUM(photo_private = 1 AND photo_quality > -1) AS private FROM `photos`  WHERE (photos.id NOT IN (SELECT photo_id FROM files WHERE file_primary = 1 AND (file_missing = 1 OR file_error <> ''))) AND (deleted_at IS NULL) LIMIT 1;
# Time: 231010 19:23:53
# User@Host: root[root] @  [10.233.105.154]
# Thread_id: 1461203  Schema: photoprism  QC_hit: No
# Query_time: 12.235543  Lock_time: 0.000031  Rows_sent: 1  Rows_examined: 489353
# Rows_affected: 0  Bytes_sent: 0
SET timestamp=1696965833;
SELECT COUNT(*) AS files FROM `files`  WHERE (file_missing = 0 AND file_root = '/' AND deleted_at IS NULL) LIMIT 1;
# User@Host: root[root] @  [10.233.105.154]
# Thread_id: 1461202  Schema: photoprism  QC_hit: No
# Query_time: 12.234036  Lock_time: 0.000125  Rows_sent: 1  Rows_examined: 489353
# Rows_affected: 0  Bytes_sent: 0
SET timestamp=1696965833;
SELECT COUNT(*) AS files FROM `files`  WHERE (file_missing = 0 AND file_root = '/' AND deleted_at IS NULL) LIMIT 1;
```

#### 6. Which software versions do you use?

(a) PhotoPrism Architecture & Build Number: AMD64, 230506-9de9a3540

(b) Database Type & Version: MariaDB

(c) Operating System Types & Versions: Linux, Docker Image

(d) Browser Types & Versions: Firefox latest, Samsung internet latest

*Always provide database and operating system details if it is a backend, import, or indexing issue. Should it be a frontend issue, at a minimum we require you to provide web browser and operating system details. When reporting a performance problem, we ask that you provide us with complete information about your environment, as there may be more than one cause.*

#### 7. On what kind of device is PhotoPrism installed?

This is especially important if you are reporting a performance, import, or indexing issue. You can skip this if you're reporting a problem you found in our public demo, or if it's a completely unrelated issue, such as incorrect page layout.

(a) Device / Processor Type: Intel Xeon E5-2670

(b) Physical Memory & Swap Space in GB: 32GB, no swap

(c) Storage Type: ZFS on NVMe, over NFS

**Comment by lastzero** [comment]
see https://github.com/photoprism/photoprism/issues/3718

**Comment by lastzero** [comment]
Also make sure to use local storage for the database instead of NFS. If it still not works then with our latest release, please open a GitHub Discussion first to determine the cause of your issues. Thank you! We have users with millions of pictures who don't have such problems.

**Comment by starcraft66** [comment]
> see #3718

I don't see anything of relevance in that issue. If you look at the slow query log I posted there are clearly multiple SQL queries that run upon logging in that take 10-20 seconds to execute. When I first installed photoprism, I had instant logins, the problem has only gotten worse over time as I add more content to my library so I doubt this has anything to do with cryptography.

**Comment by lastzero** [comment]
Looking at your logs, I see that it is a Bitnami package, while we ship the standard Mariadb image. Also, you are using an old version. Please try at least our recommended configuration and the latest release before reporting a bug. Thank you very much!

Edit: In addition, please clarify if the database also uses NFS storage or if not, what other type of storage.

**Comment by lastzero** [comment]
@starcraft66 In case you are still interested in a solution, I would appreciate it if you could test our latest preview build:

https://docs.photoprism.app/getting-started/updates/#development-preview

It includes refactored queries and trace logs that allow you to check the performance of these queries. This is what it shows for my personal library of over 100,000 files:

```
time="2023-10-15T09:45:45Z" level=trace msg="config: updated counts [223.835807ms]"
```

Although these queries are not super fast (and we know that), they should not take a minute even with 4 or 10 times the amount of pictures. That rather indicates a problem with the database configuration, or an index is broken/missing.

I'm happy to help you with troubleshooting, and hope you can forgive me for converting this into a discussion until we know what's causing it. This is our standard procedure for all problem reports. If necessary, I'll also be happy to communicate it better in the future.

**Comment by starcraft66** [comment]
Hello @lastzero, I just saw that you published a new version `231021` a couple days ago, does this release incorporate the refactored queries?

**Reply by lastzero** [reply]
@starcraft66 Yes, the update also includes the changes mentioned above. I don't expect they will make a big difference in terms of performance, but it's definitely worth testing them and you will see the exact time needed for all count queries when you enable trace log mode.

**Please also tell us more about the file system and storage technology your MariaDB database instance uses. Be aware that MariaDB 11 has been optimized for low-latency local SSD storage. So if you're using high-latency storage, better performance is to be expected with an older version.**
